[CodeGen] Introduce MachineLaneSSAUpdater for SSA Repair in Machine IR

alex-t opened the pull request:
📌 Problem
When inserting new definitions into Machine IR (e.g., reloads after spills, rematerialized values, or optimization-driven value rewriting), maintaining SSA form is non-trivial, especially with subregister lanes:

1. SSA Violation: Inserting a new def of an existing vreg creates multiple definitions, violating SSA
2. Subregister Complexity: Partial redefinitions (e.g., redefining only sub2_3 of a 128-bit register) require lane-aware PHI insertion
3. LiveInterval Corruption: Naive rewriting breaks LiveIntervals, causing downstream passes to fail
4. PHI Placement: Computing where to insert PHI nodes requires pruned Iterated Dominance Frontier (IDF) analysis
5. Use Rewriting: Rewriting uses must handle three cases:

   - Exact match: Direct substitution
   - Subset: Subreg remapping (use needs fewer lanes than def provides)
   - Super/Mixed: REG_SEQUENCE insertion (use needs more lanes than def provides)

Existing LLVM utilities (MachineSSAUpdater, LiveRangeEdit) either lack subregister awareness or are tightly coupled to specific passes, making them unsuitable for general-purpose SSA repair.

🧩 Solution
This patch introduces MachineLaneSSAUpdater, a universal SSA repair utility for Machine IR with full subregister lane awareness.
Core API:
`class MachineLaneSSAUpdater {`
`public:`
    `// Main entry point: Repair SSA for a new definition`
    `// NewDefMI: Instruction with def operand that currently defines OrigVReg (violates SSA)`
    `// Returns: Newly created virtual register`
    `Register repairSSAForNewDef(MachineInstr &NewDefMI, Register OrigVReg);`
`};`

Key Features:

1. Lane-Aware PHI Insertion: Uses pruned IDF intersected with LiveInterval subranges to place PHIs only where necessary
2. Intelligent Use Rewriting:

  - Exact match → direct substitution
  - Subset → subreg index remapping with lane mask shifting
  - Super/Mixed → REG_SEQUENCE synthesis with optimal subregister covering
3. LiveInterval Preservation: Automatically extends/recomputes LiveIntervals for all modified vregs
4. Dominance-Based Reachability: Uses MachineDominatorTree for robust def-use analysis during reconstruction
5. Non-Contiguous Lane Support: Handles complex cases like "redefine sub2 of vreg_128, leaving sub0+sub1+sub3"

🔬 Design Highlights
Algorithm Overview:
Rename: Create new vreg, update NewDefMI to define it

1. PHI Placement: Compute pruned IDF(NewDef) ∩ LiveIn(OrigVReg, DefMask)
2. Iterative PHI Insertion: Place PHIs with per-edge lane analysis
3. Use Rewriting: Walk dominated uses, apply exact/subset/super policy
4. LiveInterval Repair: Extend/recompute all affected intervals
5. Verification: Update dead flags, optionally verify MachineFunction

Pruned IDF Computation:
  - Intersects IDF of new def blocks with blocks where OrigVReg[DefMask] is live-in
  - Uses LLVM's IDFCalculatorBase for correctness
  - Avoids placing PHIs in unreachable or irrelevant blocks

REG_SEQUENCE Synthesis:

- For super-uses (use needs lanes from both old and new defs), builds REG_SEQUENCE merging:

  - Lanes from NewVReg (rewritten lanes)
  - Lanes from OrigVReg (unchanged lanes)

- Handles non-contiguous lane masks via getCoveringSubRegsForLaneMask():

  - Sorts candidates by lane count (prefer larger subregs)
  - Greedily selects minimal covering set
  - Example: LanesFromOld = 0x0F | 0xF0 → [sub0, sub2] not [lo16, hi16, sub2, sub3]


🚧 Limitations / Future Work

  - Target Scope: Currently tested on AMDGPU; other targets should work but need validation
  - Integration: Not wired into any existing passes yet (self-contained infrastructure)
  - Optimization: REG_SEQUENCE synthesis could be optimized further for specific target constraints
  - Undef Handling: Policy for undef edges (UndefEdgeAsImplicitDef) is placeholder, not fully exercised

📖 Usage Example
Spill/Reload Scenario:
`// Insert reload instruction that violates SSA`
`MachineInstr *ReloadMI = BuildMI(MBB, InsertPt, DL, TII->get(LoadOp), SpilledReg)`
                                    `.addFrameIndex(StackSlot);`

`// Repair SSA form`
`MachineLaneSSAUpdater Updater(MF, LIS, MDT, TRI);`
`Register NewReg = Updater.repairSSAForNewDef(*ReloadMI, SpilledReg);`

`// Done! ReloadMI now defines NewReg, uses are rewritten, PHIs inserted, LiveIntervals updated`

Partial Subreg Reload:
`// Reload only sub2_3 (64-bit) of a 128-bit register`
`MachineInstr *ReloadMI = BuildMI(MBB, InsertPt, DL, TII->get(LoadOp))`
                             `.addReg(SpilledReg, RegState::Define, AMDGPU::sub2_3)`
                             `.addFrameIndex(StackSlot);`

`Updater.repairSSAForNewDef(*ReloadMI, SpilledReg);`
`// Updater automatically detects lane mask from subreg index, places lane-specific PHIs`

🔗 Related Work
This patch provides the SSA repair infrastructure needed for:

  - SSA-Based Register Allocation (upcoming)
  - SSA-Based Spilling with Belady's algorithm (PR #156079 Next-Use Analysis)
  - Rematerialization optimizations
  - Late optimizations that insert new definitions

🧪 Testing
Unit Tests (llvm/unittests/CodeGen/MachineLaneSSAUpdaterTest.cpp, MachineLaneSSAUpdaterSpillReloadTest.cpp):
Test scenarios:
Simple linear reload (no PHI needed)
  ✅ Diamond CFG (PHI insertion at join point)
  ✅ Partial subreg reloads (e.g., reload only sub2_3 of vreg_128)
  ✅ Multiple predecessor PHIs with mixed lane sources
  ✅ Non-contiguous lane masks
  ✅ Same-block def-use with multiple uses
 ✅ REG_SEQUENCE synthesis for super-uses

Verification:

  - All tests run MachineFunction::verify() to ensure correctness
  - LiveInterval consistency checked via LiveInterval::verify()
  - Tests use MIR format for reproducibility

alex-t committed:
[CodeGen] Implement MachineLaneSSAUpdater for lane-aware SSA repair

This patch introduces MachineLaneSSAUpdater, a new utility for performing
SSA repair on Machine IR with full subregister lane awareness. This is
particularly important for AMDGPU and other targets with complex
subregister structures.

Key features:
- Two explicit entry points: addDefAndRepairNewDef() for new definitions
  and addDefAndRepairAfterSpill() for reload-after-spill scenarios
- Lane-aware pruned IDF computation using LLVM's IDFCalculatorBase
- Worklist-driven PHI placement algorithm that correctly handles
  iterative PHI insertion
- Per-edge lane analysis for complex PHI construction with dual-PHI
  support when both old and new register lanes are live
- SpillCutCollector helper for capturing liveness endpoints during
  spill operations with proper subrange refinement

The implementation follows standard SSA reconstruction algorithms but
extends them to handle subregister lanes properly. PHI placement uses
pruned iterated dominance frontiers, and the lane analysis ensures
correct PHI operand construction even in complex scenarios where
different lanes come from different predecessors.

The design separates concerns cleanly:
- Entry points handle scenario-specific setup (indexing, interval extension)
- Common performSSARepair() handles PHI placement and use rewriting
- Lane-aware analysis throughout maintains correctness for partial
  register operations

This is the foundation for efficient SSA repair in the presence of
complex subregister usage patterns, with rewriteDominatedUses()
implementation to follow in subsequent patches.

alex-t committed:
[CodeGen] Refactor and implement MachineLaneSSAUpdater for lane-aware SSA repair

This commit implements a comprehensive lane-aware SSA repair utility for
Machine IR with the following key changes:

- Simplify PHI creation logic: Replace complex per-edge lane analysis with
  simplified single PHI creation for reload scenarios
- Add utility function getSubRegIndexForLaneMask() for lane-to-subregister
  mapping with proper inline implementation
- Implement complete dominated use rewriting with three-case policy:
  * Exact match: direct register replacement
  * Super/Mixed: REG_SEQUENCE construction for multi-lane uses
  * Subset: preserve subregister indices for partial uses
- Add comprehensive internal helper methods:
  * incomingOnEdge() for PHI operand VNInfo analysis
  * reachedByThisVNI() for dominance and same-block ordering
  * operandLaneMask() for lane mask calculation
  * buildRSForSuperUse() for REG_SEQUENCE construction
  * extendAt() for precise LiveInterval extension
- Refactor header organization: move from forward declarations to inline
  implementations where appropriate and add missing includes

The implementation focuses on correctness for spill-reload scenarios while
maintaining lane-level precision for subregister operations.

alex-t committed:
[CodeGen] MachineLaneSSAUpdater bugfixing, refactoring and unittests. WIP.

alex-t committed:
[CodeGen] MachineLaneSSAUpdater - added new unit tests

alex-t committed:
[CodeGen] MachineLaneSSAUpdater - all unit test for all cases of the usual new definition handling finished

alex-t committed:
[CodeGen] Add MachineLaneSSAUpdater tests and simplify API

Add 11 unit tests covering SSA repair for diamonds, loops, subregisters,
and spill/reload scenarios. Fix critical bugs in subregister remapping
and LiveInterval handling.

Simplify design by removing ~250 lines of unnecessary spill-specific code
(SpillCutCollector, CutEndPoints, addDefAndRepairAfterSpill). The unified
repairSSAForNewDef() method handles all scenarios, naturally pruning
LiveIntervals through use rewriting and recomputation.

alex-t committed:
[CodeGen] MachineLaneSSAUpdater Unit test debug output fixed

vpykhtin commented:
Please, merge this into a single commit.

Add a test on a loop with single BB.

alex-t committed:
[MachineLaneSSAUpdater] Add optional NewVReg parameter to repairSSAForNewDef

This change adds an optional Register parameter to repairSSAForNewDef(),
allowing callers to provide a pre-allocated virtual register instead of
having the updater create one automatically.

Motivation:
- Enables precise control over register class selection when needed
- Required for subregister spill/reload scenarios where target-specific
  register class constraints apply (e.g., reloading a 96-bit subregister
  requires vreg_96 class, not vreg_128)
- Maintains backward compatibility - default behavior unchanged

The change is backward compatible: if NewVReg is not provided (default),
the updater creates a register automatically as before.

alex-t committed:
[MachineLaneSSAUpdater] Enhance SSA repair functionality with new methods and improved documentation.

Description::
  - Addition of `isUseReachableFromDef` methods for checking use reachability from    definitions, improving analysis efficiency.
  - Implementation of `getPrunedIDF` for computing pruned Iterated Dominance Frontiers, optimizing SSA repair processes.
  - Introduction of `insertPHIAtBlock` for explicit PHI insertion at join blocks, facilitating better control over SSA construction.
  - Refactoring of `repairSSAForNewDef` to utilize a vector for PHI register definitions, streamlining the SSA repair workflow.
  - Updates to function documentation, clarifying parameters and return values.

alex-t committed:
[MachineLaneSSAUpdater] Enhance unit tests for SSA repair methods

- Added tests for `repairSSAForNewDef` and `repairSSAForReload` to cover new functionality and ensure correct PHI insertion and use rewriting in spill/reload scenarios.
- Improved documentation within tests to clarify the behavior of SSA repair methods and their integration patterns.
- Verified that PHI operands are correctly generated and that uses are rewritten as expected in various control flow scenarios, including diamond shapes and parallel paths.
